faculty/session.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2018-2019 ASI Data Science
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from datetime import datetime, timedelta
+from collections import namedtuple
+
+import requests
+import pytz
+from six.moves import urllib
+
+import faculty.config
+
+
+AccessToken = namedtuple("AccessToken", ["token", "expires_at"])
+
+
+class MemoryAccessTokenCache(object):
+    def __init__(self):
+        self._store = {}
+
+    def get(self, profile):
+        access_token = self._store.get(profile)
+        utc_now = datetime.now(tz=pytz.utc)
+        if access_token is None or access_token.expires_at < utc_now:
+            return None
+        else:
+            return access_token
+
+    def add(self, profile, access_token):
+        self._store[profile] = access_token
+
+
+def _service_url(profile, service, endpoint=""):
+    host = "{}.{}".format(service, profile.domain)
+    url_parts = (profile.protocol, host, endpoint, None, None)
+    return urllib.parse.urlunsplit(url_parts)
+
+
+def _get_access_token(profile):
+    url = _service_url(profile, "hudson", "access_token")
+    payload = {
+        "client_id": profile.client_id,
+        "client_secret": profile.client_secret,
+        "grant_type": "client_credentials",
+    }
+
+    response = requests.post(url, json=payload)
+    response.raise_for_status()
+
+    body = response.json()
+
+    token = body["access_token"]
+    now = datetime.now(tz=pytz.utc)
+    expires_at = now + timedelta(seconds=body["expires_in"])
+
+    return AccessToken(token, expires_at)
+
+
+class Session(object):
+
+    __session_cache = {}
```

Comment:
Why is this an attribute of the class? What happens when multiple Sessions are instantiated for different profiles, do their caches point at the same object or is a new cache attached to each object? I feel it would be simpler to put it at the module level.

Reply:
I think if the factory method is on the class, the cache should be too, but maybe let's put off discussion of that until the former question is decided.